[quant][pt2] Add `move_exported_model_to_train`

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113492

Summary: This is the equivalent API to `model.train()` for 
exported models, analogous to `move_exported_model_to_eval`.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_move_exported_model_dropout
python test/test_quantization.py TestQuantizePT2E.test_move_exported_model_dropout_inplace
python test/test_quantization.py TestQuantizePT2E.test_move_exported_model_dropout_bn

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

andrewor14 committed:
[quant][pt2] Add `move_exported_model_to_train`

Summary: This is the equivalent API to `model.train()` for
exported models, analogous to `move_exported_model_to_eval`.
Note that this currently only handles dropout.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_move_exported_model_to_eval_and_train

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

leslie-fang-intel commented:
May I ask a question that what's the default behavior of dropout in a export model? Should we suggest to invoke `torch.ao.quantization.move_exported_model_to_train` after export and before `prepare_pt2e_qat`?

andrewor14 committed:
Update on "[quant][pt2] Add `move_exported_model_to_train`"


Summary: This is the equivalent API to `model.train()` for
exported models, analogous to `move_exported_model_to_eval`.
Note that this currently only handles dropout.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_move_exported_model_to_eval_and_train

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

andrewor14 commented:
> May I ask a question that what's the default behavior of dropout in a export model? Should we suggest to invoke `torch.ao.quantization.move_exported_model_to_train` after export and before `prepare_pt2e_qat`?

Hi @leslie-fang-intel, you should treat `move_exported_model_to_train` like `model.train()` for special ops (like dropout). You don't actually need to call this if the model was already exported in train mode. If the model is in eval mode (e.g. if the user called `move_exported_model_to_eval` previously) then you will have to call this function before training.

andrewor14 committed:
Update on "[quant][pt2] Add `move_exported_model_to_train`"


Summary: This is the equivalent API to `model.train()` for
exported models, analogous to `move_exported_model_to_eval`.
Note that this currently only handles dropout.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_move_exported_model_to_eval_and_train

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/pt2e/eval_utils.py:
```diff
@@ -2,19 +2,21 @@
 
 
-def _replace_dropout_for_eval(m: torch.fx.GraphModule):
+def _replace_dropout(m: torch.fx.GraphModule, for_eval_mode: bool):
```
nit: `for_eval_mode` might be a bit vague I feel, maybe just "train_to_eval" and explain in the comment saying when this is False it means eval_to_train?

andrewor14 committed:
Update on "[quant][pt2] Add `move_exported_model_to_train`"


Summary: This is the equivalent API to `model.train()` for
exported models, analogous to `move_exported_model_to_eval`.
Note that this currently only handles dropout.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_move_exported_model_to_eval_and_train

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Add `move_exported_model_to_train`"


Summary: This is the equivalent API to `model.train()` for
exported models, analogous to `move_exported_model_to_eval`.
Note that this currently only handles dropout.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_move_exported_model_to_eval_and_train

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

andrewor14 commented:
Looks like test failures are from broken trunk.

@pytorchbot merge

andrewor14 committed:
Update on "[quant][pt2] Add `move_exported_model_to_train`"


Summary: This is the equivalent API to `model.train()` for 
exported models, analogous to `move_exported_model_to_eval`.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_move_exported_model_dropout
python test/test_quantization.py TestQuantizePT2E.test_move_exported_model_dropout_inplace
python test/test_quantization.py TestQuantizePT2E.test_move_exported_model_dropout_bn

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

andrewor14 commented:
@pytorchbot merge
